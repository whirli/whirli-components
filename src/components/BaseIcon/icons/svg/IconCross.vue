<template>
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
    <path
      fill-rule="evenodd"
      clip-rule="evenodd"
      d="M21.2 4.2l-1.4-1.4-7.8 7.8-7.8-7.8-1.4 1.4 7.8 7.8-7.8 7.8 1.4 1.4 7.8-7.8 7.8 7.8 1.4-1.4-7.8-7.8z"
    />
  </svg>
</template>
