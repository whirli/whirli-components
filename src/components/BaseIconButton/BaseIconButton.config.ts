// @ts-ignore
import type { PropType } from 'vue';
import { ComponentStyles, ComponentProps } from '../../@types/components';

// Local config
import config from '@whirli-local/components/BaseIconButton/BaseIconButton.config';

// Package config
import { PropKeys, PropValues } from './BaseIconButton.constants';
import type { PropTag } from './BaseIconButton.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
};

export const ConfigProps: ComponentProps = {
  [PropKeys.IS_DISABLED]: {
    required: false,
    type: Boolean as PropType<boolean>,
    default: false,
  },
  [PropKeys.TAG]: {
    required: false,
    type: String as PropType<PropTag>,
    validator: (value: PropTag): boolean => PropValues[PropKeys.TAG].includes(value),
    default: 'button',
  },
  ...config.props,
};
