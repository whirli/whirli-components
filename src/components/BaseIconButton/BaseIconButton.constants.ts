export const PropKeys: Record<string, string> = {
  TAG: 'tag',
};

export const PropValues: Record<string, string[]> = {
  [PropKeys.TAG]: ['button', 'a'],
};
