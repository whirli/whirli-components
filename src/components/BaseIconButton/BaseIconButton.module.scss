@import '../../styles/Breakpoints.module.scss';

// default styles
.icon-button {
    background: none;
    border: none;
    box-shadow: none;
    color: inherit;
    cursor: pointer;
    display: inline-block;
}

.icon-button:disabled {
    cursor: not-allowed;
}


// breakpoint styles
$packageBreakpointStyles: ();