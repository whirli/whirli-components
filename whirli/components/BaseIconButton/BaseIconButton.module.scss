@import '@whirli-components/components/BaseIconButton/BaseIconButton.module.scss';

// Standard styles

// Breakpoint styles
$localBreakpointStyles: ();

@include createBreakpointClasses(map-merge($packageBreakpointStyles, $localBreakpointStyles));