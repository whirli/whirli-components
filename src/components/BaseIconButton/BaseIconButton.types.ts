import { PropKeys, PropValues } from './BaseIconButton.constants';
