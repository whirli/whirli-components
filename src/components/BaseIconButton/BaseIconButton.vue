<template>
  <component :is="props.tag" :class="[classes, disabledClasses]">
    <BaseIcon v-bind="$attrs" />
  </component>
</template>

<script setup lang="ts">
// Vue
import { computed, ComputedRef } from '@composition';

// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseIconButton/BaseIconButton.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseIconButton.config';
import { PropKeys } from '@whirli-components/components/BaseIconButton/BaseIconButton.constants';

// Components
import BaseIcon from '@whirli-components/components/BaseIcon/BaseIcon.vue';

// Composables
import useClasses from '@whirli-components/@use/class';

const ComponentStyles = ConfigStyles;

// @ts-ignore
const props = defineProps(ConfigProps);

// Classes
const { makeClasses } = useClasses();
const classes: string[] = [
  styles['icon-button'],
  ...makeClasses(ComponentStyles, ConfigProps, props, styles),
];
const disabledClasses: ComputedRef<string> = computed(() =>
  props[PropKeys.IS_DISABLED] ? styles['is-disabled'] : ''
);
</script>
