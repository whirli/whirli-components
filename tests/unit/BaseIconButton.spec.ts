import { mount } from '@vue/test-utils';

import BaseIconButton from '@/components/BaseIconButton/BaseIconButton.vue';
import IconCross from '@/components/BaseIcon/icons/svg/IconCross.vue';

let wrapper: any;

describe('BaseIconButton.vue - Single classes', () => {
  beforeEach(() => {
    wrapper = mount(BaseIconButton, {
      props: {
        icon: 'Cross',
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the correct icon', () => {
    expect(wrapper.findComponent(IconCross).exists()).toBeTruthy();
  });
});
