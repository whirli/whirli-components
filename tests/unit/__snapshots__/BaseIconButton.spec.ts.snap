// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseIconButton.vue - Single classes renders the component 1`] = `
<button
  class=""
  icon="Cross"
>
  <svg
    class=""
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clip-rule="evenodd"
      d="M21.2 4.2l-1.4-1.4-7.8 7.8-7.8-7.8-1.4 1.4 7.8 7.8-7.8 7.8 1.4 1.4 7.8-7.8 7.8 7.8 1.4-1.4-7.8-7.8z"
      fill-rule="evenodd"
    />
  </svg>
</button>
`;
